### PR TITLE
WIP: new recipe: lightweight CTC model for librispeech

### PR DESCRIPTION
**Pull request by wangtiance**
This PR includes:

1. A new recipe with lightweight CTC model. The encoder is similar to MobileNet V2 with a param count of just 1.76M using phone Lexicon and default model parameters, yet the WER is not far behind the TDNN-LSTM-CTC recipe. With BPE 500 lexicon the param count is 2.09M. I work on ASR for ultra-low powered edge devices, so it's interesting to see what 1-2M parameters can do. As a next step, I'll explore other lightweight models that work well with ImageNet/CIFAR-100 (such as MobileViT) and try to adapt them to ASR task.

2. changes to checkpoint.py which enables it to average state dicts containing integer data types, such as "num_batches_tracked" in nn.BatchNorm2D. Currently it throws an error. My approach is to cast it to float, do the averaging and cast it back. I'm not sure if this is a robust way.

### CODE REVIEW

File: egs/librispeech/ASR/lightweight_ctc/RESULTS.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+```bash
+python lightweight_ctc/decode.py \
+  --epoch 17 \
+  --avg 2 \
+  --use-averaged-model 1 \
+  --lang-dir data/lang_phone \
+  --exp-dir lightweight_ctc/exp_phone \
+  --max-duration 50 \
+  --num-paths 50 \
+```
```
Comment (csukuangfj) [inline egs/librispeech/ASR/lightweight_ctc/RESULTS.md:43]: Could you post the link to the tensorboard log, upload the pre-trained model to hugging face, and leave a link here to the pre-trained model?

Comment (wangtiance) [reply egs/librispeech/ASR/lightweight_ctc/RESULTS.md:43]: By pretrained model do you mean the cpu_jit.pt generated by export.py? Do I just upload it to my own huggingface? 
Also, the training log is a bit fragmented because I stopped and continued training several times (need to coordinate with other tasks). I can upload the final part.

Comment (csukuangfj) [reply egs/librispeech/ASR/lightweight_ctc/RESULTS.md:43]: > By pretrained model do you mean the cpu_jit.pt generated by export.py? Do I just upload it to my own huggingface?

Yes, `cpu_jit.pt` which is obtained from `torch.jit.script` and another file `pretrained.pt` that is by saving `model.state_dict`.

You can upload to your own account in huggingface.

---

> the training log is a bit fragmented because I stopped and continued training several times (need to coordinate with other tasks

I think tensorboard can display the logs correctly if your stop/restart the training several times.

Please also upload the decoded results, like other recipes in icefall.

File: egs/librispeech/ASR/lightweight_ctc/model.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
+import torch.nn as nn
```
Comment (csukuangfj) [inline egs/librispeech/ASR/lightweight_ctc/model.py:1]: Could you add copyright information to the beginning of this file?

File: egs/librispeech/ASR/lightweight_ctc/model.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+          expansion_rate:
+            Expansion rate of the bottleneck layers.
+          skip_add:
+            Use skip connect in the bottleneck layers.
+          rnn_dim:
+            If positive, add a GRU layer with rnn_dim as the output dimension.
+        """
+        super().__init__()
+
+        assert subsampling_factor == 4, 'Only subsampling = 4 supported.'
```
Comment (csukuangfj) [inline egs/librispeech/ASR/lightweight_ctc/model.py:45]:
```suggestion
        assert subsampling_factor == 4, 'Only subsampling = 4 is supported.', subsampling_factor
```

File: egs/librispeech/ASR/lightweight_ctc/model.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        self.bottleneck_layers = nn.Sequential()
+        for _ in range(blocks):
+            self.bottleneck_layers.append(
+                Bottleneck(
+                    in_channels=c,
+                    out_channels=c*2,
+                    expansion_rate=expansion_rate,
+                    w_stride=2,
+                    skip_add=False,
+                )),
```
Comment (csukuangfj) [inline egs/librispeech/ASR/lightweight_ctc/model.py:90]: Do you need to remove the trailing comma `,` ?

File: egs/librispeech/ASR/lightweight_ctc/model.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        Args:
+          x: input feature, with shape (N, 1, T, F),
+             where F is the feature dimension.
+
+        Returns:
+          A tensor with shape (N, round(T / 4), num_classes)
+        """
+        assert x.shape[-1] == self.num_features, \
+            f"Number of features should be {self.num_features} \
+            instead of {x.shape[3]}"
```
Comment (csukuangfj) [inline egs/librispeech/ASR/lightweight_ctc/model.py:124]:
```suggestion
            instead of {x.shape[-1]}"
```

Please be consistent.

File: egs/librispeech/ASR/lightweight_ctc/model.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        x = self.conv_subsample(x)                      # N, 32, T//4, F//4
```
Comment (csukuangfj) [inline egs/librispeech/ASR/lightweight_ctc/model.py:125]: Where is `32` from?

File: egs/librispeech/ASR/lightweight_ctc/model.py [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+        x = self.bottleneck_layers(x)                   # N, Tout, Fout * Cout
+        x = x.permute(0, 2, 3, 1).flatten(2)            # N, Tout, Fout * Cout
```
Comment (csukuangfj) [inline egs/librispeech/ASR/lightweight_ctc/model.py:127]: Could you fix the comments? The comments for the two lines are the same.

File: egs/librispeech/ASR/lightweight_ctc/model.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        in_channels: int,
+        out_channels: int,
+        expansion_rate: float = 4.0,
+        h_stride: int = 1,
+        w_stride: int = 1,
+        skip_add: bool = False,
+    ) -> None:
+        """
+        Args:
+          in_channels, out_channels: Number of input/output channels.
```
Comment (csukuangfj) [inline egs/librispeech/ASR/lightweight_ctc/model.py:151]: Please be consistent. We use the following style for docstring.

```python3
Args:
  param1:
    explanation of param1.
  param2:
    explanation of param2

Returns:
  Return xxx
```

**Comment by wangtiance** [comment]
@csukuangfj Thanks for the comments. I've made changes accordingly. Pretrained models are here: https://huggingface.co/wangtiance/lightweight_ctc

But I had trouble uploading training logs to tensorboard.dev, even with VPN on. Is there an alternative like gitee?

**Comment by csukuangfj** [comment]
> But I had trouble uploading training logs to tensorboard.dev, even with VPN on. Is there an alternative like gitee?

Could you upload training logs, `data/lang_phone`, and decoding results to hugging face?

You can use https://huggingface.co/csukuangfj/icefall-asr-librispeech-pruned-transducer-stateless5-2022-05-13/tree/main
as an example.
Screen Shot 2022-08-16 at 4 22 19 PM


I can help to upload the tensorboard logs to `tensorboard.dev` once you upload them to hugging face.

**Comment by wangtiance** [comment]
> > But I had trouble uploading training logs to tensorboard.dev, even with VPN on. Is there an alternative like gitee?
> 
> Could you upload training logs, `data/lang_phone`, and decoding results to hugging face?
> 
> You can use https://huggingface.co/csukuangfj/icefall-asr-librispeech-pruned-transducer-stateless5-2022-05-13/tree/main as an example. Screen Shot 2022-08-16 at 4 22 19 PM
> 
> I can help to upload the tensorboard logs to `tensorboard.dev` once you upload them to hugging face.

Done: https://huggingface.co/wangtiance/lightweight_ctc

File: egs/librispeech/ASR/lightweight_ctc/train.py [hunk 9]
```diff
@@ -0,0 +1,1066 @@
...
+    checkpoints = load_checkpoint_if_available(
+        params=params, model=model, model_avg=model_avg
+    )
+
+    model.to(device)
+    if world_size > 1:
+        logging.info("Using DDP")
+        model = DDP(model, device_ids=[rank])
+
+    optimizer = Eve(model.parameters(), lr=params.initial_lr)
```
Comment (pkufool) [inline egs/librispeech/ASR/lightweight_ctc/train.py:879]: From the documentation of `Eve`, it is designed to be used in `scaled` model. I don't know if `Eve` would benefit you model. You might want to try original Adam or AdamW, up to you, though.

Comment (wangtiance) [reply egs/librispeech/ASR/lightweight_ctc/train.py:879]: > From the documentation of `Eve`, it is designed to be used in `scaled` modules. I don't know if `Eve` would benefit your model. You might want to try original Adam or AdamW, up to you, though.

Thanks. I was wondering that too. Will try AdamW. Is Eden supposed to be used together with Eve?

**Comment by csukuangfj** [comment]
> Done: https://huggingface.co/wangtiance/lightweight_ctc

Thanks! I just downloaded the tensorboard logs from your repo and uploaded them to the following address:

https://tensorboard.dev/experiment/1JTkrxBrRMie3YDY3k8bYg/#scalars&_smoothingWeight=0

File: egs/librispeech/ASR/lightweight_ctc/train.py (hunk 9 above)
Comment (danpovey) [reply egs/librispeech/ASR/lightweight_ctc/train.py:879]: Yes, Eden has to be used with Eve, it is not suitable for Adam-style updates.
For a convolutional-style network with Adam, I'd probably aim for something that starts out constant and then maybe decays slowly.  E.g. halve the LR on epoch 10 and then on epoch 20, and then leave constant, I think that would be enough decrease.  The absolute value of the learning rate probably will have relatively little effect, I forget what normal values are.

Comment (wangtiance) [reply egs/librispeech/ASR/lightweight_ctc/train.py:879]: > Yes, Eden has to be used with Eve, it is not suitable for Adam-style updates. For a convolutional-style network with Adam, I'd probably aim for something that starts out constant and then maybe decays slowly. E.g. halve the LR on epoch 10 and then on epoch 20, and then leave constant, I think that would be enough decrease. The absolute value of the learning rate probably will have relatively little effect, I forget what normal values are.

Thank you. I've seen some icefall recipes using AdamW with initial lr=3e-3, and
scheduler = StepLR(optimizer, step_size=8, gamma=0.1) or 
scheduler = StepLR(optimizer, step_size=2, gamma=0.8)
The result is probably not very sensitive to initial LR given Adam is adaptive.

**Comment by csukuangfj** [comment]
Do you plan to re-open it?
@wangtiance

**Comment by wangtiance** [comment]
> 

No, I'm opening a new pull request with new model architecture and better results.

**Comment by csukuangfj** [comment]
> > 
> 
> 
> 
> No, I'm opening a new pull request with new model architecture and better results.

Thanks!
